### ISSUE TITLE
Define a Resource as a Maven dependency

### ISSUE THREAD

**Issue by rbeckman-nextgen** [nextgenhealthcare/connect#3781]
Add a new resource type to define a resource as a Maven dependency and repository.

Imported Issue. Original Details:
Jira Issue Key: MIRTH-3899
Reporter: jbartels
Created: 2016-03-08T07:12:32.000-0800

**Comment by rbeckman-nextgen** [comment]
A behavior similar to what I am after can be accomplished by taking the following manual steps:
 - create a pom see [1] for an example
 - run the maven dependencies plugin [2] {noformat}mvn org.apache.maven.plugins:maven-dependency-plugin:copy-dependencies{noformat}
 - the JARs for the dependencies in the pom and their transitive dependencies are now in ./target/dependencies

[1] example pom, this happens to pull dependencies for jasper reports:
`
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>
    <groupId>com.mirth</groupId>
    <artifactId>grab-mvn-deps</artifactId>
    <packaging>pom</packaging>
    <version>0.0.1</version>
    <dependencies>
        <dependency>
            <groupId>net.sf.jasperreports</groupId>
            <artifactId>jasperreports</artifactId>
            <version>5.6.0</version>
            <exclusions>
                <exclusion>
                    <artifactId>bcmail-jdk14</artifactId>
                    <groupId>org.bouncycastle</groupId>
                </exclusion>
                <exclusion>
                    <artifactId>bcprov-jdk14</artifactId>
                    <groupId>bouncycastle</groupId>
                </exclusion>
                <exclusion>
                    <artifactId>bcprov-jdk14</artifactId>
                    <groupId>org.bouncycastle</groupId>
                </exclusion>
                <exclusion>
                    <artifactId>bcmail-jdk14</artifactId>
                    <groupId>bouncycastle</groupId>
                </exclusion>
                <exclusion>
                    <artifactId>jdtcore</artifactId>
                    <groupId>eclipse</groupId>
                </exclusion>
                <exclusion>
                    <artifactId>log4j</artifactId>
                    <groupId>log4j</groupId>
                </exclusion>
                <exclusion>
                    <groupId>org.slf4j</groupId>
                    <artifactId>slf4j-log4j12</artifactId>
                </exclusion>
                <exclusion>
                    <groupId>org.slf4j</groupId>
                    <artifactId>jcl104-over-slf4j</artifactId>
                </exclusion>
            </exclusions>            
        </dependency>  
        <dependency>
            <groupId>net.sf.jasperreports</groupId>
            <artifactId>jasperreports-fonts</artifactId>
            <version>4.0.0</version>
        </dependency>     
    </dependencies>
</project>
`

[2] - https://maven.apache.org/plugins/maven-dependency-plugin/copy-dependencies-mojo.html

Imported Comment. Original Details:
Author: jbartels
Created: 2016-03-08T08:28:43.000-0800

**Comment by rbeckman-nextgen** [comment]
We would not do anything like calling "mvn" in a local OS shell from within the MC JVM... Instead, there would have to be a way to download the JAR every time the resource is updated. Remember that all this needs to be dynamic and multiple-classloader friendly; it's not something that would be done only at build time or server startup time.

There's probably some embedded way to do parse a POM, and get download URLs for each dependency ID. Note that this would require that the server have access to the internet, which is not always possible depending on the environment.

Imported Comment. Original Details:
Author: narupley
Created: 2016-03-08T08:39:56.000-0800

**Comment by rbeckman-nextgen** [comment]
A similar need came up on another project and I was able to do some research.

I was able to find some additional options:
 - http://maven.apache.org/ref/3-LATEST/maven-embedder/project-summary.html
 - http://maven.apache.org/shared/maven-invoker/
 - https://github.com/TimMoore/mojo-executor

Maven Invoker looks really good, but under the covers it just makes syscalls to "mvn". 

Mojo Executor is not actively maintained, but it does appear to be pure Java. I'm not able to find it making syscalls. I think it directly invokes the Maven Mojo classes. What is less clear is how to load its execution environment.


Imported Comment. Original Details:
Author: jbartels
Created: 2016-11-18T14:43:16.000-0800
